src/Lecture4.hs

```diff
@@ -134,8 +140,21 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+split :: Eq a => a -> [a] -> [[a]]
+split sep list = case elemIndex sep list of
+    Just i -> let (xs, ys) = splitAt i list
+        in xs : split sep (drop 1 ys)
+    Nothing -> [list]
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow s = case split ',' s of
+    _:_:_:_:_ -> Nothing
+    s1:s2:s3:_ -> (if null s1 then Nothing else Just s1)
+        >>= \v1 -> readMaybe s2
+        >>= \v2 -> readMaybe s3
+        >>= \t3 -> if t3 < 0 then Nothing else Just t3
+        >>= \v3 -> Just (Row v1 v2 v3)
+    _ -> Nothing
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow s = do
    [v1, s2, s3] <- Just (split ',' s)
    guard (v1 /= "")
    v2 <- readMaybe s2
    v3 <- readMaybe s3
    guard (v3 >= 0)
    Just (Row v1 v2 v3)
```